Adds emphasis to Select-String default formatter

derek-xia opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary
Adds the -Emphasize parameter for Select-String and makes it work with -AllMatches and -SimpleMatch.
The -Emphasize parameter highlights the matched text using the negative escape sequence. Issue #2905.
Started at HackIllinois.
<!-- Summarize your PR between here and the checklist. -->

## PR Context
Makes it easier to distinguish what users are looking for when using the Select-String cmdlet. Helps resolve issue #2905.

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/4479
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` to your commit messages if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

Commit:
[feature] Added Select-String -Color and made it work with -AllMatch and -SimpleMatch

Commit:
[feature] Added Select-String -Color and test cases

Commit:
[feature] Adds -Emphasize for Select-String

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=8963) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=8963) before we can accept your contribution.<br/><br/>:x: derek-xia [sign now](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=8963)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/PowerShell/PowerShell?pullRequest=8963) it.</sub>

rkeithhill commented:
I would suggest that you check to see if the host support virtual terminal sequences, the use a VT sequence instead of asterisks.  Using any character is a bit problematic because you may have that character in the text your searching.  That could confuse the user.  FWIW, this is how grep "highlights" matched text:

![image](https://user-images.githubusercontent.com/5177512/53293618-82c2c200-3794-11e9-9bfa-63537a8cf721.png)

The tricky part is figuring out what color (or effect e.g. reverse) to use for matched text.  And whatever you pick, you probably need to make it configurable.

TylerLeonhardt commented:
Thanks @rkeithhill for the guidance 😊

This is a student at [HackIllinois](https://www.hackillinois.org). @rjmholt and I are working with the students here and are reviewing PRs that they send.

All comments are welcome 😊

rkeithhill commented:
FYI I find this VT sequence resource very handy: https://docs.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences

vexx32 commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -1614,12 +1681,10 @@ private bool DoMatchWorker(string operandString, MatchInfo matchInfo, out MatchI
-                                  Line = operandString,
-                                  Pattern = Pattern[patternIndex]
-                              };
+                matchResult = new MatchInfo(false, indexes, lengths, Emphasize);
```
```suggestion
                matchResult = new MatchInfo(false, indexes, lengths, Emphasize.IsPresent);
```

vexx32 reviewed (commented):
I also think that highlighting with some form of colour is a more worthwhile approach, as the matched string itself could easily contain asterisks or really any number of other characters that we might try to emphasize it with.

Commit:
[feature] Adds red highlighting for matched text

Commit:
[feature] Adds red highlighting for matched text

derek-xia commented:
I made the matched text red on terminals that support VT sequences and retained the asterisks on terminals that don't support VT sequences.

TylerLeonhardt commented:
@derek-xia can you add a screenshot so everyone can see?

derek-xia commented:
<img width="1082" alt="screen shot 2019-02-24 at 1 56 15 am" src="https://user-images.githubusercontent.com/17694694/53296590-8af02100-37d7-11e9-9d74-05e7a7e81372.png">
This is a screenshot of -Emphasize with and without -AllMatch

rjmholt commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -93,6 +93,50 @@ public class MatchInfo
+        /// <summary>
+        /// Stores the starting index of each match within the line.
+        /// </summary>
+        private readonly List<int> matchIndexes;
```
```suggestion
        private readonly List<int> _matchIndexes;
```

rjmholt commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -93,6 +93,50 @@ public class MatchInfo
+        /// <summary>
+        /// Stores the length of each match within the line.
+        /// </summary>
+        private readonly List<int> matchLengths;
```
```suggestion
        private readonly List<int> _matchLengths;
```

rjmholt commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -93,6 +93,50 @@ public class MatchInfo
+        /// <summary>
+        /// Stores a values indicating whether or not the terminal supports VT.
+        /// </summary>
+        private readonly bool? supportsVirtualTerminal;
```
```suggestion
        private readonly bool? _supportsVirtualTerminal;
```

rjmholt replied:
Rather than make this field nullable, it should be `false` instead of `null` in all instances I think

rjmholt commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -1518,6 +1600,9 @@ private bool DoMatchWorker(string operandString, MatchInfo matchInfo, out MatchI
             int patternIndex = 0;
             matchResult = null;
 
+            List<int> indexes = new List<int>();
```
```suggestion
            var indexes = new List<int>();
```

rjmholt commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -1518,6 +1600,9 @@ private bool DoMatchWorker(string operandString, MatchInfo matchInfo, out MatchI
             matchResult = null;
 
+            List<int> indexes = new List<int>();
+            List<int> lengths = new List<int>();
```
```suggestion
            var lengths = new List<int>();
```

rjmholt commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1:
```diff
@@ -71,6 +71,30 @@ Describe "Select-String" -Tags "CI" {
 	}
 
+	it "Should return an array of matching strings with the first match highlighted when Emphasize is used" {
+		If ($IsWindows) {
```
```suggestion
		if ($IsWindows)
		{
```

rjmholt commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1:
```diff
@@ -71,6 +71,30 @@ Describe "Select-String" -Tags "CI" {
+	it "Should return an array of matching strings with the first match highlighted when Emphasize is used" {
+		If ($IsWindows) {
+			$testinputone | Select-String -Pattern "l" -Emphasize | Should -Be "he*l*lo", "he*l*lo"
+		} Else {
```
```suggestion
		}
		else
		{
```

rjmholt replied:
Ditto below

rjmholt committed:
Apply suggestions from code review

Co-Authored-By: derek-xia <derek.xia@gmail.com>

Commit:
[feature] Refactored code

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -221,12 +265,43 @@ public override string ToString()
+                {
+                    int offset = open.Length + close.Length;
+                    sb.Insert(_matchIndexes[i] + (i * offset), open);
+                    sb.Insert(_matchIndexes[i] + _matchLengths[i] + ((i * offset) + open.Length), close);
```
The algorithm is very expensive. I'd use String.Create() because we can calculate target string length.

derek-xia replied:
Done.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -1518,6 +1600,9 @@ private bool DoMatchWorker(string operandString, MatchInfo matchInfo, out MatchI
             matchResult = null;
 
+            var indexes = new List<int>();
+            var lengths = new List<int>();
```
I wonder why we need this if already have the same in matchResult.Matches?
Note that in both cases we have disordered list.

derek-xia replied:
I'm creating new lists in all cases, although I could use matchResult.Matches in most cases and allocate and use these lists only when the SimpleMatch parameter is enabled. If the SimpleMatch parameter is enabled, then the current behavior is that Matches does not contain any Match objects. Without any Match objects, I would need to pass in these lists to implement the colored text when SimpleMatch is enabled.

rjmholt replied:
I think we should avoid creating or populating lists in the (possibly common) case where `Emphasize` is not specified or where VT100 isn't available. We don't want a performance hit in cases where there's no benefit.

That might look something like:

```csharp
List<int> indexes;
List<int> lengths;

if (Emphasize)
{
    indexes = new List<int>();
    lengths = new List<int>();
}

...

if (Emphasize)
{
    foreach (Match match in matches)
    {
        indexes.Add(match.Index);
        lengths.Add(match.Length);
    }
}

// etc.

...
    matchResult = Emphasize
        ? new MatchInfo(indexes, lenghts)
        : new MatchInfo();

    matchResult.IgnoreCase = !CaseSensitive;
    Line = operandString;
    // etc.
...
```

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1:
```diff
@@ -71,6 +71,33 @@ Describe "Select-String" -Tags "CI" {
 	}
 
+	it "Should return an array of matching strings with the first match highlighted when Emphasize is used" {
+		if ($IsWindows)
```
Doesn't  Windows support VT?

TylerLeonhardt replied:
Yeah rather than doing `$IsWindows` you can do:

```
$Host.UI.SupportsVirtualTerminal
```

Which returns a bool.

rkeithhill replied:
The first version of Windows 10 (1507) did not support VT. Probably not many folks still using that version but $Host.UI.SupportsVirtualTerminal just as well.  And you may be running in a different host than conhost.

felixfbecker replied:
But those hosts would just not display the color, right?

Would also be best to consider https://no-color.org/

felixfbecker commented:
Does it highlight capture group like in https://github.com/PowerShell/PowerShell/issues/2905?

felixfbecker commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1:
```diff
@@ -71,6 +71,33 @@ Describe "Select-String" -Tags "CI" {
+		}
+		else
+		{
+			$testinputone | Select-String -Pattern "l" -Emphasize | Should -Be "he[31ml[0mlo", "he[31ml[0mlo"
```
Shouldn't there be an ESC byte before each `[`?
```suggestion
			$testinputone | Select-String -Pattern "l" -Emphasize | Should -Be "he`e[31ml`e[0mlo", "he`e[31ml`e[0mlo"
```

derek-xia replied:
Thanks, I added the ESC byte.

felixfbecker commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1:
```diff
@@ -71,6 +71,33 @@ Describe "Select-String" -Tags "CI" {
+	it "Should return an array of matching strings with the first match highlighted when Emphasize is used" {
+		if ($IsWindows)
+		{
+			$testinputone | Select-String -Pattern "l" -Emphasize | Should -Be "he*l*lo", "he*l*lo"
```
Imo emphasis is a formatting concern. The output of the cmdlet should be the raw data, and any highlighting should be done through PowerShell's formatting system (i.e. script blocks in `Format.ps1xml` files), so that piping to other cmdlets is still possible.

Examples:
https://github.com/felixfbecker/PowerGit#log
https://github.com/felixfbecker/PowerGit/tree/master/PowerGit/Formats

vexx32 commented:
@joeyaiello mentioned on Twitter that might be worth considering just making this behaviour the default. Perhaps that might be worth doing? If there are concerns with stability / how different conhosts would work with this we can always stick it behind an experimental feature flag. @TylerLeonhardt, what do you think?

TylerLeonhardt commented:
@vexx32, I think @felixfbecker brought up a very good point - this is more of a formatting concern.

Looking at the formatting code, it seems that formatting `MatchInfo` to be printed to the console is [done here](https://github.com/PowerShell/PowerShell/blob/d80154430d600a52a1d1e3a9f3ecf032b36185d4/src/System.Management.Automation/FormatAndOutput/DefaultFormatters/PowerShellCore_format_ps1xml.cs#L358-L366). Where it simply runs:

```powershell
$_.ToString(((get-location).path))
```

where `$_` is the instance of `MatchInfo`

All it's doing is calling the `ToString`. Perhaps what would be better is move @derek-xia's logic added to `ToString` to a new public method on `MatchInfo` called `Emphasize` and then the formatting line linked above could just call:

```powershell
$_.Emphasize(((get-location).path))
```

Does that sound alright @vexx32 @felixfbecker?

rkeithhill commented:
Or maybe call it `ToAnsiString()`.

iSazonov commented:
Considering that we would like to receive color input for any output type, this should not be in every single cmdlet, it should be out.

powercode commented:
I don't think this change should not be made in Select-String. 

You can check out this gist that adds coloring of the matchinfo output as a custom format. https://gist.github.com/powercode/4833804efd23045387bd5d5249d76f7b 

There already exists an issue for improving the formatting of MatchInfo, but this is a bit of "*nix" thinking where flags on a command modify the output. Make it a view in the formatting system instead.

rjmholt commented:
I feel like we're all agreeing here. Just want to summarise:

- This behaviour should be in the formatter
- Most efficient implementation is probably as @rkeithhill suggests with a `ToAnsiString()` method on `MatchInfo` objects. Possibly `ToEmphasizedString()` if we want that to return something still useful on a platform that does not support ANSI/VT control codes (like emphasis characters around the match).
- Given that it would be in the formatter, we could have it on by default (can changing formatting be a breaking change?)

My only question is: should there be a switch for this on the cmdlet?

vexx32 commented:
If I recall correctly, changes to formatting are not considered "breaking."

Emphasis _characters_ are kind of hard to argue for, really. There's no guarantee such characters won't already be somewhere in or around the matched string, unfortunately. Might still be worth doing, I suppose, but honestly I feel simply omitting the formatting where it's not supported is likely to be most effective.

I don't think there need be a switch for it. Maybe define one formatview for emphasized and one for non-emphasized?

iSazonov commented:
I like the coloring but I am not a fan of turning PowerShell into an artist. We do not embed a code formatter, but simply publish AST API. The same with coloring. Perhaps it is not formatting system api but conhost api.
Turn on/off:
```powershell
$PSColoring = $true # Globally
Get-Date | Format-List -Coloring # Locally
```

rjmholt commented:
@iSazonov I agree with you in principle, but I think this kind of match emphasis brings PowerShell up to speed with `grep` in stakes that PowerShell is usually better at: interactive experience.

If we can get this into the formatter, then we can remove or refactor it later with no obligation not to break things.

So making the suggested change in the immediate gets us what we want, without committing us to anything down the road (except a public API on `MatchInfo` that we shouldn't guarantee dependable behaviour on).

rjmholt commented:
> ```powershell
> $PSColoring = $true # Globally
> Get-Date | Format-List -Coloring # Locally
> ```

Maybe worth writing up an RFC for this?

iSazonov commented:
> we can have something small in the meantime too.

We already have the "small" - it is Jason's PSCore. Implementing vNext formatting system based on classes simplify adding new feature such as coloring (new interface?). So I thing right way is implementing PSCore project first. (Before Coloring RFC)

felixfbecker commented:
Advanced output formatting is and has always been part of PowerShell - if your shell is object based, and you also want to provide a strong interactive experience, you need a good formatting system, that supports tables, nested data, grouping, truncation, and coloring too. But all of this is separate from the data and API. It shouldn't be exposed in a parameter switch, nor in the returned strings.

I like the idea of respecting the user's choice regarding not wanting color output. However I feel like PowerShell has the opportunity here to handle that on a shell level? E.g. it could define a well-known global variable `$PSEnableColors` that if set to `$false`, will make PowerShell trim all color escape sequences from formatting output, so that not each cmdlet has to deal with it on their own.

TylerLeonhardt commented:
@iSazonov... Do you mean https://github.com/lzybkr/PSMore

Not sure what the status is on that project... (cc @lzybkr)


I want to also remind everyone here that some folks that submit PRs do not know everything about PowerShell and we should treat that as a learning experience to teach folks about PowerShell-specific concepts.

powercode commented:
I think the jury is still out on how do handle colors. I have started on an implementation for better coloring of select-string, but deferred it until we know what to do with colors. That opens up a can of worms, with theming etc. (which may be very important for the color-blind).

vexx32 commented:
If we assign a default "emphasis" color or two from the default console colors, they can theme it themselves?

iSazonov commented:
@TylerLeonhardt Yes, PSMore. You can find more information in #7857 and referenced there issues. I think it has great potential but it costs a lot to start.

TylerLeonhardt commented:
Don't want to lose this awesome feature to stalebot - I think this discussion has given us direction:

1. we should "solve" colors in Formatting all up... This is a substantial change that we expect to see with adoption of PSMore eventually or by some other contribution to PowerShell... however that's out of scope for this PR
2. @derek-xia can still move the changes he made to `MatchInfo.ToString()` to a new function called `ToEmphasizedString()` (or similar) and then [change the formatting](https://github.com/PowerShell/PowerShell/blob/d80154430d600a52a1d1e3a9f3ecf032b36185d4/src/System.Management.Automation/FormatAndOutput/DefaultFormatters/PowerShellCore_format_ps1xml.cs#L358-L366) for `MatchInfo` to use that method instead of `ToString`

iSazonov commented:
I don't think it's worth doing anything other than PSMore.

vexx32 commented:
Even with PSMore, surely `MatchInfo` would need some kind of underlying API to call upon to figure out where to apply colors?

And given that PSMore at least on the surface of what we can see appears to be a fairly low-priority project for the PS Team, I think there is some value in improving the base formatting subsystem where we can in small ways while we wait for PSMore to be more thoroughly fleshed out. 🙂

iSazonov commented:
Current FS is very power but not designed for coloring, paging and so on.
I believe any attempts to stick them in FS only add complexity and confusion.
PSMore idea is just to add new capabilities and features without changing source types and objects.

felixfbecker commented:
There is nothing lacking in the formatting system that would prevent this feature and therefor this feature should not be blocked on PSMore.

lzybkr commented:
I haven't touched any of my side projects for 2+ months, let alone PSMore. My day job is very far removed from PowerShell, so unless somebody picks up my ideas from PSMore - I'd discourage waiting on me.

iSazonov commented:
@lzybkr Can we move the PSMore project to public under PowerShell org?
/cc @SteveL-MSFT

SteveL-MSFT commented:
@iSazonov I would be fine moving PSMore to PowerShell org, but it's up to @lzybkr.  Also keep in mind I currently don't have any resources to apply to that project.

iSazonov commented:
@SteveL-MSFT I remember about limited resources. I believe, in any case, we need platforms/repos for an alpha code that can fire in a year or two.

TylerLeonhardt commented:
@derek-xia consider [what I said above](https://github.com/PowerShell/PowerShell/pull/8963#issuecomment-469668360) as the current plan for this PR... if you're still interested in finishing up the contribution!

derek-xia commented:
Thanks for the comments, I'm definitely still interested in finishing up the contribution! I'll work on it sometime today or tomorrow.

Commit:
[feature] Moved new functionality, changed formatter, fixed tests

Commit:
[feature] Fixed style errors

powercode commented:
Hard coding colors in the only formatter can't be our solution. Pwsh is used with different backgrounds etc, so we need a better solution. I argue for solving this with scripts until we have a more solid solution for colors. Some support for themed colors would be really useful.

felixfbecker commented:
@powercode so is `grep`, and it colors its output. This shouldn't be a reason why we can't add nice features.

powercode commented:
I think we should try to solve the underlying problem.

vexx32 commented:
Hard coding colours can be the immediate solution, to be updated when we figure out theming. 

Theming is tricky business, although perhaps @jaykul would be interested in offering some pointers from his Pansies module and Powerline.

powercode commented:
I've poked him before :) Time to do it again, right @Jaykul? :)

daxian-dbw commented:
I'm also concerned that hard coding the emphasizing color in the default view might cause that default view not usable for a user who changes the console background color.
I know formatting change is not consider a breaking change, but only if it doesn't affect usability.

IIRC, the formatting system will use the first view definition for a type as its default view, and there is no way to change the default unless you override the default view with another formatting definition.
So, for users that use a different background color, they might have to create a `ps1xml` file to override the default, which is very unfriendly in my opinion.

TylerLeonhardt commented:
@daxian-dbw or they do:

```
Select-String | % ToString
```

One of the things mentioned above by @vexx32  was using a preexisting console color like:

```
$host.PrivateData.WarningForegroundColor
```

Then that way, if the user happens to change the color for readability purposes then that changes with it.

TylerLeonhardt commented:
Just to remind everyone how nice this looks.

![image](https://user-images.githubusercontent.com/2644648/54093933-2518a300-435a-11e9-98d3-acfbdf2149e5.png)

TylerLeonhardt commented:
I just thought of something... instead of color... how about the ANSI escape sequence for _underline_?

![image](https://user-images.githubusercontent.com/2644648/54094725-32d12700-4360-11e9-9fd3-45498287ab59.png)

It doesn't go down the rabbit hole of color, but it still stands out.

vexx32 commented:
Are there any console / terminal compat issues with underline? I wouldn't expect there should be, but I don't see it used much, so I suppose it should be asked. 😄

TylerLeonhardt commented:
I initially thought about blink 
```powershell
"`e[5mhello world`e[0m" 
```
* note the 5 instead of 4

but that seems to only be supported in Terminal.app on macOS (instead of iTerm which is what I use). Case by case but at least there's no color concerns.

daxian-dbw commented:
@TylerLeonhardt Thanks, missed that comment. Thanks @vexx32, I didn't know it previously :)

Then we probably have a better solution -- to make the color we use change automatically based on the `Console.BackgroundColor`.
We could have a switch to map the `ConsoleColor` members (background color) to the escape sequences to use every time `ToEmphasizedString` is invoked. So the emphasize color can change appropriately when the background color is changed.

vexx32 commented:
At first glance, I might think perhaps a dictionary would be a more extensible solution looking forwards... but I don't really know how that might pan out. 😄

daxian-dbw commented:
@vexx32 Agreed. It looks using the escape sequence of the color from `$host.PrivateData.WarningForegroundColor` might be the best temporary solution before we have a more universal and extensible way of doing this in the formatting system.

BTW, I think `Yellow` (default warning fore color) as the default for emphasizing is pretty eye-catching too :)

daxian-dbw commented:
To summarize, I think the proposal is to have a public method in `MatchInfo` type as in below (no other changes in `Select-String`):
```
public string ToEmphasizedString(string directory, ConsoleColor? emphasizingColor)
{
    if (emphasizingColor == null)
    {
        return ToString(directory);
    }

    string open, close;
    switch (emphasizingColor.Value)
    {
        // Map every member to the corresponding escape sequences
        case ConsoleColor.Black:
            open = "\u001b[30m";
            close = "\u001b[0m";
            break;
        case ConsoleColor.DarkBlue:
             ...
    }

    // Actual logic of the method starts
    ...
}
```
In the format definition, do this:
```
.AddScriptBlockExpressionBinding(@"
    $path =(Get-Location).Path
    $color = $Host.PrivateData.WarningForegroundColor -as [System.ConsoleColor]
    $_.ToEmphasizedString($path, $color)")
```

@vexx32 @derek-xia Please take a look to see if there is any flaws.

vexx32 commented:
The immediate concern I have there @daxian-dbw is that only one of those colours is configurable.

In windows, you can modify the RGB value of any console colour via the Properties / Defaults dialog screen accessed from a console or a shortcut to a console application. You can't do the same thing with an escape sequence.

If, for example, a user reconfigures their console to use primarily black text on white background by doing so, the console has no way of knowing this is being done (that I can find). This means a simple "this colour is **named** black, so we should use a light colour" (etc) solution may end up making things _harder_ to read instead of easier.

That said... Given this behaviour, I don't see any easy way to rectify this at all, unless there is some way to pull the current RGB value of the console colours being referenced.

vexx32 commented:
@daxian-dbw that looks pretty solid! I guess the only potential pitfalls there is that (on Windows, at least) console colours aren't a fixed colour value. Do those escape sequences refer to specific _console_ colours or are they fixed RGB values?

daxian-dbw commented:
> Do those escape sequences refer to specific console colours or are they fixed RGB values

Not sure if I understand you question correctly. In windows, we can see the fixed RGB values for each of the member of `ConsoleColor` in the console 'Property'->'Colors' (from left to right: Black, DarkBlue, DarkGreen ... Yellow, White). Can't we get the escape sequences based on those RGB values?

![image](https://user-images.githubusercontent.com/127450/54165338-7e95d600-441d-11e9-9ac6-ea7a748ec097.png)

vexx32 commented:
Yes, that's kind of what I was asking. I haven't seen a way to get at the RGB values. As long as we can use those we should be good.

lzybkr commented:
There is an API on Windows to get the RGB values of the 16 console colors, see [here](https://github.com/lzybkr/PSReadLine/blob/a878b3686cba23ad39fcd98a6b418c9db0bb77ab/PSReadLine/ScreenCapture.cs#L245).

Based on [these docs](https://invisible-island.net/xterm/ctlseqs/ctlseqs.html), it looks like you could use `ESC 4 ; <c> ; ? BEL` (replacing <c> with the color entry you want) - but this sequence doesn't seem to work on Windows.

Jaykul commented:
@daxian-dbw In my opinion, you *should not* use $Host.PrivateData (or you should have a fallback option). Those colors only exist in the default host, and you're not actually warning, anyway.

Personally, I would vote for adding a `$PSColors` dictionary preference variable (which we will eventually add more things to, I have a list) with `EmphasisColors` as an array of ConsoleColor names. If you only take _one_ emphasis color, you're going to be loosing functionality by comparison to [the formatting example](https://gist.github.com/powercode/4833804efd23045387bd5d5249d76f7b) linked in [@powercode's comment](#issuecomment-466960651).

@lzybkr the [OSC 4 query stuff is not implemented in Windows yet](https://github.com/Microsoft/console/issues/317), and it's also not implemented in lots of other terminals. I don't think that would be safe to use -- but honestly @vexx32 there's no reason to use full RGB escape sequences. 

ConsoleColor should be fine until such time as we add a RGB Color class _throughout_ the shell.

iSazonov commented:
I still wonder why we change a code if coloring must be in formatting system as @powercode mentioned above https://github.com/PowerShell/PowerShell/pull/8963#issuecomment-466960651

powercode commented:
Will this method be a part of the public API?

Or can it be removed when we come up with better solutions?

daxian-dbw commented:
> I still wonder why we change a code if coloring must be in formatting system as @powercode mentioned above #8963 (comment)

@iSazonov Even after the formatting system is updated to support color configuration, I'm afraid a method like this will still be needed just like how the method `ToString(string directory)` is needed today for the formatting of `MatchInfo` object. But it would be must more intuitive to configure the emphasizing color.

> Will this method be a part of the public API? Or can it be removed when we come up with better solutions?

@powercode It will be easier to use in the formatting script if we make it public. We can add the `ObsoleteAttribute` to the method when it's time to remove it.
Or, if we want no leftover at all, we can make that method internal, and call it using reflection in the formatting script. Then nothing is public and can be removed/changed at anytime.

---------------------------------------

We all agree that the ultimate solution should be done in the formatting system -- make it support color formatting in a uniform and extensible way.
What was proposed in https://github.com/PowerShell/PowerShell/pull/8963#issuecomment-471770352 is to have a **temporary solution** to colorize the matching text produced from `Select-String`. It's by no means the long-term thing, but may be good to have to make this particular interactive experience just work for most PowerShell users.

This PR is put in the `6.2.0-GA-Consider` milestone. If we can agree on a temp solution that is low risk and easy to revert in future, then we can take it to the 6.2.0 GA. If not, I'm fine that we continue to override the default view with custom formatting provided by @lzybkr and @powercode before a long-term solution is done in the formatting system.

iSazonov commented:
I see this idea with coloring of Select-String output very desirable. In this case, I would still choose the path with the script and overriding the default view as a temporary solution.

Jaykul commented:
Either negative or underline would work well, I think. I mean, it's not as cool as the color-per-group thing, but they should be readable regardless of people's actual colors, without borrowing the WarningColor from PrivateData.

TylerLeonhardt commented:
@derek-xia still interested in getting this in? Based on the feedback, we should change the escape character to the Negative escape sequence.

Commit:
[feature] Changed red escape sequence to negative escape sequence

Commit:
[feature] Removed surrounding asterisks if VT is unsupported

derek-xia commented:
I changed the escape character to the negative escape sequence. Here's what it looks like on a terminals with light and dark backgrounds.
<img width="846" alt="powershellwhite" src="https://user-images.githubusercontent.com/17694694/57007131-db0aab80-6bab-11e9-910e-750802d9faac.png">
<img width="809" alt="powershellgreen" src="https://user-images.githubusercontent.com/17694694/57007134-e958c780-6bab-11e9-9c71-9edf6b7b4128.png">
I also removed the asterisks that would previously surround the selected text on terminals that don't support virtual terminal escape sequences. I can see why they would be confusing, unless there's a better character to use.

Jaykul commented:
Looks good to me

TylerLeonhardt commented:
We will need to rebase this PR. @derek-xia can you go ahead and do that?

This seems like a good guide:
https://github.com/servo/servo/wiki/Beginner's-guide-to-rebasing-and-squashing

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -93,6 +93,49 @@ public class MatchInfo
+        /// Gets or sets a value indicating whether the matched portion of the string is highlighted.
+        /// </summary>
+        /// <value>Whether the matched portion of the string is highlighted with asterisks.</value>
+        public bool Emphasize { get; set; }
```
I believe this should be private. We don't use it publicly and have no need.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -93,6 +93,49 @@ public class MatchInfo
+        /// <summary>
+        /// Stores a values indicating whether or not the terminal supports VT.
+        /// </summary>
+        private readonly bool _supportsVirtualTerminal;
```
I believe we should remove this from MatchInfo. This is only used in `ToEmphasizedString()`
where we can directly use `Host.UI.SupportsVirtualTerminal`.

derek-xia replied:
I don't think I can remove this easily. MatchInfo doesn't have access to `Host`, while SelectStringCommand does since it inherits from PSCmdlet.

derek-xia replied:
I moved this check into SelectStringCommand's doMatchWorker.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -93,6 +93,49 @@ public class MatchInfo
+        /// Gets or sets a value indicating whether the matched portion of the string is highlighted.
+        /// </summary>
+        /// <value>Whether the matched portion of the string is highlighted with asterisks.</value>
+        public bool Emphasize { get; set; }
```
I believe this should be private. We don't use it publicly and have no need.

derek-xia replied:
Thanks, fixed.

Commit:
[feature] Changed MatchInfo's Emphasize to private

derek-xia commented:
I finished rebasing the PR. Should I squash the commits as well?

rjmholt commented:
> I finished rebasing the PR. Should I squash the commits as well?

No need for that -- the commits are squashed on merge

rjmholt commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -93,6 +93,49 @@ public class MatchInfo
+        /// <param name="matchLengths">Sets the matchLengths.</param>
+        /// <param name="emphasize">Used for implementing -Emphasize.</param>
+        /// <param name="vt">Sets a value indicating whether or not virtual terminal is supported.</param>
+        public MatchInfo(List<int> matchIndexes, List<int> matchLengths, bool emphasize, bool vt)
```
The `emphasize` and `vt` parameters seem unneeded. Instead I think there should be two constructors:
- The 0-parameter constructor means there's no emphasis (could privately have an `_emphasize` variable set to `false` for clarity).
- The 2-parameter constructor that takes indices and lengths means there is emphasis (could privately have `_emphasize` set to `true` but could also just check `_matchIndexes != null && _matchLengths != null`.

derek-xia replied:
Sounds good. I'll add a check for whether VT is supported before the 2-parameter constructor is used since there's otherwise no way to check that within matchInfo.

rjmholt reviewed (commented):
Looks really good @derek-xia. I've left a couple of comments about some slight changes that won't affect functionality at all (just performance mainly)

rjmholt commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -248,6 +291,74 @@ public string ToString(string directory)
+                for (int i = 0; i < _matchIndexes.Count; i++) 
+                {
+                    // Adds characters before match
+                    while (lineIndex < _matchIndexes[i])
```
```suggestion
                    Line.CopyTo(lineIndex, chars, charsIndex, _matchIndexes[i] - lineIndex);
                    charsIndex += _matchIndexes[i] - lineIndex;
                    lineIndex = _matchIndexes[i];
```

derek-xia replied:
I'm getting "cannot convert from 'System.Span<char>' to 'char[]'" when I use CopyTo.

rjmholt replied:
Ah I looked for this comment but couldn't find it (got a notification but it didn't take me to the comment).

This `string.Create` method is fairly obscure -- the documentation on it isn't great.

I'd go with array creation; .NET will probably be somewhat faster using `CopyTo` rather than while loops

iSazonov replied:
CopyTo() works for Span-s too https://docs.microsoft.com/en-us/dotnet/api/system.readonlyspan-1.copyto?view=netstandard-2.1

like
```c#
Line.AsSpan.Slice(...).CopyTo(chars.Slice(...));
```

rjmholt replied:
Ah I searched for that -- didn't think to look for properties. My problem with `string.Create(...)` here is that it creates the variable `buf` and I don't know what the type is

iSazonov replied:
Really Line is mapped to buf in the context.
Direction is that .Net Core is more and more optimized for Span and ReadOnlySpan. The fact is that PowerShell Core consumes more memory than Windows PowerShell. So we would use new patterns everywhere.

rjmholt replied:
Ah that's right -- `buf` if the span of the old string and `chars` is the span of the new one.

But that's unclear at best. As much as I agree about using spans over strings, that API is very poorly documented ([`String.Create`](https://docs.microsoft.com/en-us/dotnet/api/system.string.create?view=netcore-3.0#System_String_Create__1_System_Int32___0_System_Buffers_SpanAction_System_Char___0__) and the needed [`SpanAction<T, TArg>`](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.spanaction-2?view=netcore-3.0) doc &mdash; the only example usage I've found is [in a blog post](https://msdn.microsoft.com/en-us/magazine/mt814808.aspx), so I feel .NET could endorse this direction a little bit better) and in this case I don't think it really buys us all that much. Plus we allocate a closure.

We can always refactor it later.

iSazonov replied:
Best docs is Core itself https://source.dot.net/#System.Private.CoreLib/shared/System/String.cs,7dd9a20e8a84bf21,references

rjmholt commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -248,6 +291,74 @@ public string ToString(string directory)
+                    }
+                    
+                    // Adds opening vt sequence
+                    for (int j = 0; j < open.Length; j++)
```
```suggestion
                   open.CopyTo(0, chars, charsIndex, open.Length);
                   charsIndex += open.Length;
```

rjmholt commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -248,6 +291,74 @@ public string ToString(string directory)
+                    }
+                    
+                    // Adds characters being emphasized
+                    for (int j = 0; j < _matchLengths[i]; j++)
```
```suggestion
                    Line.CopyTo(lineIndex, chars, charsIndex, _matchLengths[i]);
                    lineIndex += _matchLengths[i];
                    charIndex += _matchLengths[i];
```

rjmholt commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -248,6 +291,74 @@ public string ToString(string directory)
+                    }
+
+                    // Adds closing vt sequence
+                    for (int j = 0; j < close.Length; j++)
```
```suggestion
                    close.CopyTo(0, chars, charsIndex, close.Length);
                    charsIndex += close.Length;
```

rjmholt commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -248,6 +291,74 @@ public string ToString(string directory)
+                }
+
+                // Adds remaining characters in line
+                while (lineIndex < Line.Length)
```
```suggestion
                Line.CopyTo(lineIndex, chars, charIndex, Line.Length - lineIndex);
```

rjmholt commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -248,6 +291,74 @@ public string ToString(string directory)
+        /// <returns>The colored string representation of the match object.</returns>
+        public string ToEmphasizedString(string directory)
+        {
+            if (!Emphasize || ! _supportsVirtualTerminal) {
```
```suggestion
            if (_matchIndexes == null || _matchLengths == null) {
```

rjmholt replied:
Or
```suggestion
            if (!_emphasize) {
```

rjmholt commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -248,6 +291,74 @@ public string ToString(string directory)
+            string open = "\u001b[7m";
+            string close = "\u001b[0m";
+
+            Line = string.Create((_matchIndexes.Count * (open.Length + close.Length)) + Line.Length, Line, (chars, buf) =>
```
Or see my comment below for another alternative

rjmholt commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -248,6 +291,74 @@ public string ToString(string directory)
+                }
+            });
+
+            string modifiedLine = ToString(directory);
```
The way `Line` is used here is a bit opaque I think. It might be that a better approach is to create a private overload of `ToString()` and put all the current `ToString(directory)` logic into it like:

```csharp
public string ToString(string directory)
{
    return ToString(Line, directory);
}

private string ToString(string line, string directory)
{
    ... // actual content with 'Line' changed to 'line'
}
```

Then for the method you'd want:
```csharp
public string ToEmphasizedString(string directory)
{
    if (!_emphasize)
    {
        return ToString(directory);
    }

    string emphasizedLine = EmphasizeLine();
    return ToString(emphasizedLine, directory);
}

private string EmphasizeLine()
{
    const string invertColorsVT100 = "\u001b[7m";
    const string resetVT100 = "\u001b[0m";
    return string.Create((_matchIndexes.Count * (open.Length + close.Length)) + Line.Length, Line, (chars, buf) => {
        // Complicated logic
    });
}
```

derek-xia committed:
[feature] Cleaned up matchInfo and DoMatchWorker

derek-xia committed:
[feature] Changed ToEmphasizedString string creation

derek-xia committed:
[feature] Refactored ToEmphasizedString

rjmholt commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -1518,6 +1628,18 @@ private bool DoMatchWorker(string operandString, MatchInfo matchInfo, out MatchI
+            // If Emphasize is set and VT is supported,
+            // the lengths and starting indexes of regex matches
+            // need to be passed in to the matchInfo object.
+            if (Emphasize && Host.UI.SupportsVirtualTerminal)
```
Given how many times `Emphasize && Host.UI.SupportsVirtualTerminal` it might be worth saving its result for reuse:

```csharp
bool shouldEmphasize = Emphasize && Host.UI.SupportsVirtualTerminal;

if (shouldEmphasize)
{
    ...
}
```

rjmholt reviewed (approved):
Looks good -- left a small comment

TylerLeonhardt reviewed (approved):
LGTM (besides @rjmholt's comment)!! So excited to get this in.

One thing we'll need is updated docs... I just wrote up how to do that. Go head and follow this:

https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#quick-steps-if-youre-changing-an-existing-cmdlet

and let me know how it goes. Once this gets merged in and that doc PR gets merged in, you'll be able to do:

```powershell
Update-Help
Get-Help Select-String
```

and see your changes there :)

TravisEz13 committed:
Update Select-String.Tests.ps1

TylerLeonhardt committed:
readd curly brace

vexx32 commented:
Did we ever get a clear decision as to the necessity of the `-Emphasize` parameter in the cmdlet @TylerLeonhardt?

Formatting changes rarely require new parameters, and while I see a lot of discussion it doesn't seem to have gone anywhere. If anything, I'd be looking at `-NoEmphasize` being the new parameter (with the default being to show emphasis). We've gotten past the colour-accessibility concerns by using the inverted terminal sequence rather than a specific color, so is there still any reason this shouldn't be default behaviour?

derek-xia committed:
shouldEmphasize

derek-xia committed:
Formatting changes

SteveL-MSFT commented:
@vexx32 we'll discuss it in the next @PowerShell/powershell-committee meeting.  I tend to agree with you on this one.

iSazonov commented:
@SteveL-MSFT If your team will discuss the cmdlet please look my RFC - I want to continue and make the cmdlet more fast and power power.

joeyaiello commented:
@iSazonov I think we reviewed that one on Monday, and I hadn't even seen your comment! :) 

And yeah, I'm leaning hard towards the `-NoEmphasis` parameter. As @rjmholt said way above, this is basically table stakes for `grep`, and as a formatting change it's not breaking. I get that some people are text purists, but this is crazy helpful to most people (myself included), and I don't think it's particularly jarring.

SteveL-MSFT commented:
@PowerShell/powershell-committee reviewed this.  We agree that emphasis should be default and there should be a `-NoEmphasis` switch.  There is some concern about the tests not being complete for TERMs that don't support VT100 escape codes, but agree this is out of scope for this PR and a general issue we should discuss in the future.

TylerLeonhardt commented:
@derek-xia do you feel comfortable with this change of making your emphasis logic the default behavior and changing the switch to `-NoEmphasis`?

joeyaiello commented:
FYI, we've met with accessibility folks, and the current "highlighted text" approach to the emphasis is acceptable.

derek-xia commented:
@TylerLeonhardt Yeah, I think it makes sense. I'll work on that.

derek-xia committed:
Changed Emphasize to NoEmphasis

rjmholt reviewed (approved):
LGTM

SteveL-MSFT commented:
Looks like the last issue is that the test needs to be fixed

TylerLeonhardt committed:
add right }

TylerLeonhardt committed:
tab instead of spaces?

TylerLeonhardt commented:
there was a missing brace so I quickly added it :) lets see what CI says but this should be good to go.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1:
```diff
@@ -127,6 +127,7 @@ Describe "Select-String" -Tags "CI" {
 
 	It "Should return ParameterBindingException when -Raw and -Quiet are used together" {
 		{ $testinputone | Select-String -Pattern "hello" -Raw -Quiet -ErrorAction Stop } | Should -Throw -ExceptionType ([System.Management.Automation.ParameterBindingException])
+        }
```
This brace seems out of place?  I think this whole file should be reformatted so the intention is correct making it easier to understand the braces.

vexx32 replied:
Maybe that would be better done in a quick follow-up PR so we can keep track of the diffs and so accidental changes are made?

SteveL-MSFT replied:
The reformat can be a single commit.  Then the fix of the brace a separate commit.

TylerLeonhardt replied:
That test isn't touched in this PR. My last commit just undid an accidental deletion of a brace.

SteveL-MSFT replied:
@TylerLeonhardt perhaps you can just add a commit after reformatting the tests?

TylerLeonhardt replied:
What's wrong with the formatting exactly? I don't see anything wrong with the changes in this PR.

SteveL-MSFT replied:
Not this PR specifically, but if you [view the whole file](https://github.com/PowerShell/PowerShell/blob/46e62fd2dd88f36b765c194a67cdcf18049b715e/test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1), the indentation is all over the place.  Makes it hard to review the code.  I still can't tell how that missing brace fixed the code since the indentation isn't consistent.

vexx32 replied:
The indentation issue is because about half the file is indented with tabs instead of spaces. Github renders a tab as 8 spaces, whereas most editors commonly only render it with 4 spaces (though this is configurable).

I think VS Code has a convert tabs to spaces option in its command palette to get rid of the tabs, and then you can do a quick skim over the file to fix any remaining issues.

SteveL-MSFT replied:
Ok, I see.  For this repo, we've standardized on spaces.

TylerLeonhardt replied:
done.

TylerLeonhardt committed:
tabs to spaces

TylerLeonhardt commented:
@SteveL-MSFT please rereview

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1:
```diff
@@ -4,245 +4,246 @@
 Describe "Select-String" -Tags "CI" {
     $nl = [Environment]::NewLine
```
line 5 and 6 should really be in a `BeforeAll{}`

TylerLeonhardt replied:
added

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1:
```diff
@@ -4,245 +4,246 @@
-	it "Should be called without errors" {
-	    { $testinputone | Select-String -Pattern "hello" } | Should -Not -Throw
-	}
+            it "Should return an array data type when multiple matches are found" {
```
lines 16 to 19 are indented over one too many

TylerLeonhardt replied:
fixed

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1:
```diff
@@ -4,245 +4,246 @@
+            Select-String second $testInputFile -Raw -Context 2,2 | Should -BeExactly $expected
+        }
     }
     Push-Location $currentDirectory
```
This should be in a `AfterAll{}` and at the top, not at the bottom

TylerLeonhardt replied:
added

SteveL-MSFT reviewed (approved):
@TylerLeonhardt I realize my feedback is in the existing code and not part of this specific PR.  So if you want to fix it as part of this PR, that's fine, or if you want to submit a new PR that's fine too.

TylerLeonhardt committed:
address steve's comments

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -243,6 +293,61 @@ public string ToString(string directory)
+        /// Surrounds the matched text with virtual terminal sequences to invert it's color. Used in ToEmphasizedString.
+        /// </summary>
+        /// <returns>The matched line with matched text inverted.</returns>
+        private string EmphasizeLine()
```
Good candidate to be static.

TylerLeonhardt replied:
But it uses `_matchIndexes`?

iSazonov replied:
We can pass `_matchIndexes` by parameter.

rjmholt replied:
I agree. Just like a whole bunch of other PowerShell code that is similarly caught up by stateful object state implicitly passed in to methods.

But I believe that pre-existed here and this contribution shouldn't be held back on the basis of code debt incurred before PowerShell was open-sourced. This PR has been open long enough and has been a good, responsive, constrained contribution and I think we really need to just accept or reject it, rather than turning it into a Ship of Theseus.

vexx32 replied:
Yeah, I think this one's gone on long enough and is a definite nice-to-have in PSv7 😁

iSazonov replied:
The PR is delayed because it is in new area for PowerShell. However, the code should be clean.

SteveL-MSFT replied:
Since this isn't a public API, we can always refactor internal code.  I would suggest a separate PR if it's a concern or @iSazonov you can always add a commit to this PR if you feel strongly about this one.

iSazonov replied:
Yes, I can :-) Do you ok that we have over 70 opened PR? :-)

SteveL-MSFT replied:
@iSazonov I'll ask the team to start driving down the number of open PRs

adityapatwardhan commented:
@derek-xia Thank you for your contribution!

TylerLeonhardt commented:
AT LAST! Well done @derek-xia 🎉

TylerLeonhardt commented:
I'm just gonna leave this here...

https://twitter.com/fatherjack/status/1199807445037723648?s=19

😁
